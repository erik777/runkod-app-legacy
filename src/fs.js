import {PATH_SEPARATOR, BASE_PATH} from './constants';


const arrToPath = (arr) => `${PATH_SEPARATOR}${arr.join(PATH_SEPARATOR)}${PATH_SEPARATOR}`;

const pathToArr = (path) => path.split(PATH_SEPARATOR).filter((x) => x !== '');

const isPath = (st) => st.startsWith(BASE_PATH);

const generatePaths = (path) => {
  /*
  Generates all possible paths from a path

  Example:
    "/static/media/fonts/"
    returns:
    "/static/",
    "/static/media/",
    "/static/media/fonts/"
  */

  const paths = pathToArr(path);

  // Walk on the array and return paths consecutively
  return paths.map((x, i) => {
    const p = paths.slice(0, i + 1);
    return arrToPath(p);
  });
};

const buildPathMap = (files) => {
  let paths = [BASE_PATH];

  files.forEach(x => {
    paths = [...paths, ...generatePaths(x.parent).filter(x => !paths.includes(x))]
  });

  const m = {};

  paths.forEach(x => {
    m[x] = {'folders': [], 'files': []};

    m[x]['folders'] = paths.filter(y => {
      if (y.indexOf(x) !== 0) {
        return false;
      }

      return y.replace(x).split('/').length === 2
    });

    m[x]['files'] = files.filter(y => {
      return y.parent === x
    });
  });

  return m;
};

const pathLabel = (path) => {
  /*
  Returns label of a path

  Example:
    "/static/media/fonts/"
    returns:
    fonts
  */
  return pathToArr(path).slice(-1);
};

const parentPath = (path) => {
  /*
    Returns parent path of a path

    Example:
      "/static/media/fonts/"
      returns:
      /static/media/
    */

  if (path === BASE_PATH) {
    return null
  }

  const parentArr = pathToArr(path).slice(0, -1);
  return parentArr.length === 0 ? BASE_PATH : arrToPath(parentArr);
};

const pathWalk = (map, path, rv) => {
  for (let folder of map[path]['folders']) { // eslint-disable-line no-unused-vars
    rv.push(folder);
    pathWalk(map, folder, rv)
  }
};


const getFilesUnderPath = (map, path) => {
  /*
    Returns all files under a path with a dict list.

    Example output:
      [{
        "name": "index.json",
        "parent": "/cars/",
      },
       {
        "name": "amg.json",
        "parent": "/cars/sport/mercedes/",
      },
      {
        "name": "s300.json",
        "parent": "/cars/sport/mercedes/",
      }]
   */

  let folders = [path];
  let files = [];
  pathWalk(map, path, folders);

  folders.forEach(x => {
    files = [...files, ...map[x]['files']]
  });

  return files
};

export default {
  arrToPath,
  pathToArr,
  isPath,
  generatePaths,
  buildPathMap,
  pathLabel,
  parentPath,
  pathWalk,
  getFilesUnderPath
};