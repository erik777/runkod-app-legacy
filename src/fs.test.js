import fs from "./fs"

it('1- generate paths', () => {
  const input = '/static/media/fonts/';
  const result = fs.generatePaths(input);
  expect(result).toMatchSnapshot();
});

it('2- generate paths', () => {
  const input = '/';
  const result = fs.generatePaths(input);
  expect(result).toMatchSnapshot();
});


it('3- generate paths', () => {
  const input = '/images/';
  const result = fs.generatePaths(input);
  expect(result).toMatchSnapshot();
});


it('4- buildPathMap', () => {
  const files = [
    {
      parent: '/',
      name: 'asset-manifest.json'
    },
    {
      parent: '/',
      name: 'index.html'
    },
    {
      parent: '/',
      name: 'manifest.json'
    },
    {
      parent: '/',
      name: 'precache-manifest.cbe4dd161c6842da089c5aea3c10c7de.js'
    },
    {
      parent: '/',
      name: 'service-worker.js'
    },
    {
      parent: '/images/',
      name: 'icon.png'
    },
    {
      parent: '/images/',
      name: 'og.jpg'
    },
    {
      parent: '/static/css/',
      name: '2.9a3ef949.chunk.css'
    },
    {
      parent: '/static/css/',
      name: '2.9a3ef949.chunk.css.map'
    },
    {
      parent: '/static/css/',
      name: 'main.61b1d037.chunk.css'
    },
    {
      parent: '/static/css/',
      name: 'main.61b1d037.chunk.css.map'
    },
    {
      parent: '/static/js/',
      name: '2.30431126.chunk.js'
    },
    {
      parent: '/static/js/',
      name: '2.30431126.chunk.js.map'
    },
    {
      parent: '/static/js/',
      name: 'main.8c32f0c4.chunk.js'
    },
    {
      parent: '/static/js/',
      name: 'main.8c32f0c4.chunk.js.map'
    },
    {
      parent: '/static/js/',
      name: 'runtime~main.c5541365.js'
    },
    {
      parent: '/static/js/',
      name: 'runtime~main.c5541365.js.map'
    },
    {
      parent: '/static/media/fonts/',
      name: 'm-plus-rounded-1c-latin-100.b9c1b5c3.woff'
    }
  ];

  const result = fs.buildPathMap(files);
  expect(result).toMatchSnapshot();
});


it('5- buildPathMap 2', () => {
  const files = [
    {
      parent: '/images/portrait/2019/06/',
      name: 'alice.png'
    },
    {
      parent: '/images/portrait/2019/07/',
      name: 'bob.png'
    },
    {
      parent: '/cv/2019/06/',
      name: 'alice.pdf'
    },
    {
      parent: '/cv/2019/07/',
      name: 'bob.pdf'
    }
  ];

  const result = fs.buildPathMap(files);
  expect(result).toMatchSnapshot();
});


it('6- path label', () => {
  const path = '/static/media/fonts/';

  const result = fs.pathLabel(path);
  expect(result).toMatchSnapshot();
});


it('7- path label', () => {
  const path = '/images/';

  const result = fs.pathLabel(path);
  expect(result).toMatchSnapshot();
});


it('8- parent path', () => {
  const path = '/static/media/fonts/foo/bar';

  const result = fs.parentPath(path);
  expect(result).toMatchSnapshot();
});


it('9- parent path', () => {
  const path = '/static/';

  const result = fs.parentPath(path);
  expect(result).toMatchSnapshot();
});


it('10- parent path', () => {
  const path = '/';

  const result = fs.parentPath(path);
  expect(result).toMatchSnapshot();
});


it('11- getFilesUnderPath', () => {

  const files = [
    {
      parent: '/cars/',
      name: 'index.json'
    },
    {
      parent: '/cars/sport/mercedes/',
      name: 'amg.json'
    },
    {
      parent: '/cars/sport/mercedes/',
      name: 's300.json'
    },
    {
      parent: '/cars/sport/mercedes/',
      name: 'e180.json'
    },

    {
      parent: '/cars/sport/bmw/',
      name: 'm3.json'
    },
    {
      parent: '/cars/sport/bmw/',
      name: 'm5.json'
    },
    {
      parent: '/bikes/sport/yamaha/',
      name: 'f1.json'
    },
    {
      parent: '/bikes/naked/yamaha/',
      name: 'fz8.json'
    }
  ];

  const map = fs.buildPathMap(files);

  let rv = [];
  const result = fs.getFilesUnderPath(map, '/cars/', rv);
  expect(result).toMatchSnapshot();


  const result2 = fs.getFilesUnderPath(map, '/cars/sport/mercedes/', rv);
  expect(result2).toMatchSnapshot();


  const result3 = fs.getFilesUnderPath(map, '/cars/sport/bmw/', rv);
  expect(result3).toMatchSnapshot();

  const result4 = fs.getFilesUnderPath(map, '/', rv);
  expect(result4).toMatchSnapshot();
});


it('11- isPath', () => {
  expect(fs.isPath('arandomtext')).toMatchSnapshot();
  expect(fs.isPath('/cars/')).toMatchSnapshot();
});