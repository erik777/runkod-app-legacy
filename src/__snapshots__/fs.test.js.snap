// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`1- generate paths 1`] = `
Array [
  "/static/",
  "/static/media/",
  "/static/media/fonts/",
]
`;

exports[`2- generate paths 1`] = `Array []`;

exports[`3- generate paths 1`] = `
Array [
  "/images/",
]
`;

exports[`4- buildPathMap 1`] = `
Object {
  "/": Object {
    "files": Array [
      Object {
        "name": "asset-manifest.json",
        "parent": "/",
      },
      Object {
        "name": "index.html",
        "parent": "/",
      },
      Object {
        "name": "manifest.json",
        "parent": "/",
      },
      Object {
        "name": "precache-manifest.cbe4dd161c6842da089c5aea3c10c7de.js",
        "parent": "/",
      },
      Object {
        "name": "service-worker.js",
        "parent": "/",
      },
    ],
    "folders": Array [
      "/images/",
      "/static/",
    ],
  },
  "/images/": Object {
    "files": Array [
      Object {
        "name": "icon.png",
        "parent": "/images/",
      },
      Object {
        "name": "og.jpg",
        "parent": "/images/",
      },
    ],
    "folders": Array [],
  },
  "/static/": Object {
    "files": Array [],
    "folders": Array [
      "/static/css/",
      "/static/js/",
      "/static/media/",
    ],
  },
  "/static/css/": Object {
    "files": Array [
      Object {
        "name": "2.9a3ef949.chunk.css",
        "parent": "/static/css/",
      },
      Object {
        "name": "2.9a3ef949.chunk.css.map",
        "parent": "/static/css/",
      },
      Object {
        "name": "main.61b1d037.chunk.css",
        "parent": "/static/css/",
      },
      Object {
        "name": "main.61b1d037.chunk.css.map",
        "parent": "/static/css/",
      },
    ],
    "folders": Array [],
  },
  "/static/js/": Object {
    "files": Array [
      Object {
        "name": "2.30431126.chunk.js",
        "parent": "/static/js/",
      },
      Object {
        "name": "2.30431126.chunk.js.map",
        "parent": "/static/js/",
      },
      Object {
        "name": "main.8c32f0c4.chunk.js",
        "parent": "/static/js/",
      },
      Object {
        "name": "main.8c32f0c4.chunk.js.map",
        "parent": "/static/js/",
      },
      Object {
        "name": "runtime~main.c5541365.js",
        "parent": "/static/js/",
      },
      Object {
        "name": "runtime~main.c5541365.js.map",
        "parent": "/static/js/",
      },
    ],
    "folders": Array [],
  },
  "/static/media/": Object {
    "files": Array [],
    "folders": Array [
      "/static/media/fonts/",
    ],
  },
  "/static/media/fonts/": Object {
    "files": Array [
      Object {
        "name": "m-plus-rounded-1c-latin-100.b9c1b5c3.woff",
        "parent": "/static/media/fonts/",
      },
    ],
    "folders": Array [],
  },
}
`;

exports[`5- buildPathMap 2 1`] = `
Object {
  "/": Object {
    "files": Array [],
    "folders": Array [
      "/images/",
      "/cv/",
    ],
  },
  "/cv/": Object {
    "files": Array [],
    "folders": Array [
      "/cv/2019/",
    ],
  },
  "/cv/2019/": Object {
    "files": Array [],
    "folders": Array [
      "/cv/2019/06/",
      "/cv/2019/07/",
    ],
  },
  "/cv/2019/06/": Object {
    "files": Array [
      Object {
        "name": "alice.pdf",
        "parent": "/cv/2019/06/",
      },
    ],
    "folders": Array [],
  },
  "/cv/2019/07/": Object {
    "files": Array [
      Object {
        "name": "bob.pdf",
        "parent": "/cv/2019/07/",
      },
    ],
    "folders": Array [],
  },
  "/images/": Object {
    "files": Array [],
    "folders": Array [
      "/images/portrait/",
    ],
  },
  "/images/portrait/": Object {
    "files": Array [],
    "folders": Array [
      "/images/portrait/2019/",
    ],
  },
  "/images/portrait/2019/": Object {
    "files": Array [],
    "folders": Array [
      "/images/portrait/2019/06/",
      "/images/portrait/2019/07/",
    ],
  },
  "/images/portrait/2019/06/": Object {
    "files": Array [
      Object {
        "name": "alice.png",
        "parent": "/images/portrait/2019/06/",
      },
    ],
    "folders": Array [],
  },
  "/images/portrait/2019/07/": Object {
    "files": Array [
      Object {
        "name": "bob.png",
        "parent": "/images/portrait/2019/07/",
      },
    ],
    "folders": Array [],
  },
}
`;

exports[`6- path label 1`] = `
Array [
  "fonts",
]
`;

exports[`7- path label 1`] = `
Array [
  "images",
]
`;

exports[`8- parent path 1`] = `"/static/media/fonts/foo/"`;

exports[`9- parent path 1`] = `"/"`;

exports[`10- parent path 1`] = `null`;

exports[`11- getFilesUnderPath 1`] = `
Array [
  Object {
    "name": "index.json",
    "parent": "/cars/",
  },
  Object {
    "name": "amg.json",
    "parent": "/cars/sport/mercedes/",
  },
  Object {
    "name": "s300.json",
    "parent": "/cars/sport/mercedes/",
  },
  Object {
    "name": "e180.json",
    "parent": "/cars/sport/mercedes/",
  },
  Object {
    "name": "m3.json",
    "parent": "/cars/sport/bmw/",
  },
  Object {
    "name": "m5.json",
    "parent": "/cars/sport/bmw/",
  },
]
`;

exports[`11- getFilesUnderPath 2`] = `
Array [
  Object {
    "name": "amg.json",
    "parent": "/cars/sport/mercedes/",
  },
  Object {
    "name": "s300.json",
    "parent": "/cars/sport/mercedes/",
  },
  Object {
    "name": "e180.json",
    "parent": "/cars/sport/mercedes/",
  },
]
`;

exports[`11- getFilesUnderPath 3`] = `
Array [
  Object {
    "name": "m3.json",
    "parent": "/cars/sport/bmw/",
  },
  Object {
    "name": "m5.json",
    "parent": "/cars/sport/bmw/",
  },
]
`;

exports[`11- getFilesUnderPath 4`] = `
Array [
  Object {
    "name": "index.json",
    "parent": "/cars/",
  },
  Object {
    "name": "amg.json",
    "parent": "/cars/sport/mercedes/",
  },
  Object {
    "name": "s300.json",
    "parent": "/cars/sport/mercedes/",
  },
  Object {
    "name": "e180.json",
    "parent": "/cars/sport/mercedes/",
  },
  Object {
    "name": "m3.json",
    "parent": "/cars/sport/bmw/",
  },
  Object {
    "name": "m5.json",
    "parent": "/cars/sport/bmw/",
  },
  Object {
    "name": "f1.json",
    "parent": "/bikes/sport/yamaha/",
  },
  Object {
    "name": "fz8.json",
    "parent": "/bikes/naked/yamaha/",
  },
]
`;
